cli/doctest_runner.rs

```diff
@@ -0,0 +1,223 @@
+use regex::Regex;
+use std::path::{Path, PathBuf};
+
+use crate::fs as deno_fs;
+use crate::installer::is_remote_url;
+use crate::test_runner::is_supported;
+
+pub struct DocTest {
+  // This removes repetition of imports in a file
+  imports: std::collections::HashSet<String>,
+  // This contains codes in an @example section with their imports removed
+  bodies: Vec<DocTestBody>,
+}
+
+struct DocTestBody {
+  caption: String,
+  line_number: usize,
+  path: PathBuf,
+  value: String,
+  ignore: bool,
+  is_async: bool,
+}
+
+pub fn prepare_doctest(
+  include: Vec<String>,
+  // root_path: &PathBuf,
+) -> Vec<DocTest> {
+  let include_paths: Vec<_> =
+    include.into_iter().filter(|n| !is_remote_url(n)).collect();
+
+  let mut prepared = vec![];
+
+  for path in include_paths {
+    let p = deno_fs::normalize_path(&Path::new(&path));
+    if p.is_dir() {
+      let test_files = crate::fs::files_in_subtree(p, |p| {
+        let supported_files = ["ts", "tsx", "js", "jsx"];
+        match p.extension().and_then(std::ffi::OsStr::to_str) {
+          Some(x) => supported_files.contains(&x) && !is_supported(p),
+          _ => false,
+        }
+      });
+      prepared.extend(test_files);
+    } else {
+      prepared.push(p);
+    }
+  }
+
+  prepared
+    .iter()
+    .filter_map(|dir| {
+      // TODO(iykekings) use deno error instead
+      let content = std::fs::read_to_string(&dir).expect(
+        format!("File doesn't exist {}", dir.to_str().unwrap_or("")).as_str(),
+      );
+      extract_jsdoc_examples(content, dir.to_owned())
+    })
+    .collect::<Vec<_>>()
+}
+
+fn extract_jsdoc_examples(input: String, p: PathBuf) -> Option<DocTest> {
+  lazy_static! {
+    static ref JS_DOC_PATTERN: Regex =
+      Regex::new(r"/\*\*\s*\n([^\*]|\*[^/])*\*/").unwrap();
+    // IMPORT_PATTERN doesn't match dynamic imports
+    static ref IMPORT_PATTERN: Regex =
+      Regex::new(r"import[^(].*\n").unwrap();
+    static ref EXAMPLE_PATTERN: Regex = Regex::new(r"@example\s*(?:<\w+>.*</\w+>)*\n(?:\s*\*\s*\n*)*```").unwrap();
+    static ref TICKS_OR_IMPORT_PATTERN: Regex = Regex::new(r"(?:import[^(].*)|(?:```\w*)").unwrap();
+    static ref CAPTION_PATTERN: Regex = Regex::new(r"<caption>([\s\w\W]+)</caption>").unwrap();
+    static ref TEST_TAG_PATTERN: Regex = Regex::new(r"@example\s*(?:<\w+>.*</\w+>)*\n(?:\s*\*\s*\n*)*```(\w+)").unwrap();
+    static ref AWAIT_PATTERN: Regex = Regex::new(r"\Wawait\s").unwrap();
+  }
+
+  let mut import_set = std::collections::HashSet::new();
```

Comment:
you use this full path multiple times in this file, should probably be switched to a `use` statement.